docs/architecture/app-class-and-template.md

```diff
@@ -0,0 +1,160 @@
+# ApplicationClass and ApplicationTemplate
+
+Managing [Applications](./../terminology.md#Application) can be passed on some external system, to remove the burden of integration with Compass from Applications.
+Because of that [IntegrationSystem](./../terminology.md#Integration-System) and ApplicationClass types were introduced.
+
+To increase the reusability of the application configuration, we introduce ApplicationTemplate.
+The diagram below describes when to use ApplicationClass or ApplicationTemplate.
+
+![](./assets/app-vs-template-vs-class.svg)
+
+## Terminology
+
+- **ApplicationClass** represents an object for creating Applications. When you create an application from it, the IntegrationSystem
+dynamically populates definition.
+   
+- **ApplicationTemplate** represents an object for creating Applications from the static definition.
+
+## GraphQL API Changes
+   
+1. Managing Application Class
+
+```
+input ApplicationClassInput {
+    name: String!
+    description: String
+    global: Boolean! = true
+    documents: [Document!]
+    integrationSystemID: ID!
+    previewApiSpecs: [ApiSpecInput!]
+    previewEventApiSpecs: [EventApiSpecInput!]
+}
+
+input ApplicationInput {
+    ...
+    applicationClassName: String
+    ...
+}
+
+
+type Mutation {
+    createApplicationClass(in: ApplicationClassInput!): ApplicationClass!
+    updateApplicationClass(id: ID!, in: ApplicationClassInput): ApplicationClass!
+    deleteApplicationClass(id: ID!): ApplicationCLass!
+}
+
+type Query {
+    applicationClasses(first: Int = 100, after: PageCursor): ApplicationClassPage!
+    applicationClass(id: ID): ApplicationClass!
+}
+
+```
+`ApplicationClass.documents` is a place for documenting any manual steps required for managing Application of given class by 
+a Integration System defined by `integrationSystemID`. For example, it can contain information on how to register a given application
+in the integration system if it is not performed by the Integration System automatically. 
+
+`previewApiSpecs` and `previewEventApiSpecs` are optional fields that document what kind of API a user can expect after creating 
+Application originated from the ApplicationClass. The final API can differ from those presented in preview fields, for example, some API can be missing
+because the user lacks privileges to access them. 
+
+In the first iteration, ApplicationClass can be registered only globally (`ApplicationClass.global`), so it will be available for every tenant. 
+To create the Application of given class, the user set field `applicationClassName` in `ApplicationInput`. This field is optional.
+In case `applicationClassName` is provided, Compass adds two labels to the Application:
+- `applicationClass`
+- `inegrationSystemID`
+
```

Comment:
Also, please mention that some fields provided by user may be overwritten  by Integration System. For example, if user provides some APIs for Application managed by Integration System, Integration System should update the list with correct APIs. I'm missing explicit piece of information that some Application fields are controlled by Integration System and any user input will be overwritten by Integration System, which has to watch for changes and reset user changes for App.

Reply:
I rephrased ApplicationClass definition:
`When you create an application from it, the IntegrationSystem dynamically updates Application definition, such us APIs, EventAPIs etc.`
>  I'm missing explicit piece of information that some Application fields are controlled by Integration System and any user input will be overwritten by Integration System, which has to watch for changes and reset user changes for App.

Currently, we don't have a clear concept about that, so at the beginning probably IntegrationSystem can update all fields.  Also, the behavior of IntegrationSystem is undefined if there are some APIs already registered manually.